pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -4,33 +4,147 @@ package nodemaintenance
+	drain "k8s.io/kubectl/pkg/drain"
+)
+
+var _ reconcile.Reconciler = &ReconcileNodeMaintenance{}
+
+const (
+	NodeMaintenanceSpecAnnotation   = "lifecycle.openshift.io/maintenance"
+	NodeMaintenanceStatusAnnotation = "lifecycle.openshift.io/maintenance-status"
+	LeaseHolderIdentity             = "node-maintenance"
+	LeasePaddingSeconds             = 30
+	WaitForLeasePeriod              = 10 * time.Second
+)
+
+const (
+	MinMaintenenceWindowSeconds = 300              // minimum maintenance window size. (must be bigger than LeasePaddingSeconds)
+	EvictionTimeSlice           = 10 * time.Second // max. time in seconds that eviction may block.
+	RequeuDrainingWaitTime      = 10 * time.Second // time to reqeueu draining if not completed.
+)
+
+type LeaseStatus int
+
+const (
+	LeaseStatusNotFound               LeaseStatus = -1 // currently not active lease
+	LeaseStatusOwnedByMe              LeaseStatus = 1  // there is a lease active right now
+	LeaseStatusOwnedByDifferentHolder LeaseStatus = 2
+	LeaseStatusFail                   LeaseStatus = 3
+)
+
+type NodeMaintenanceStatusType string
+
+const (
+	NodeStateWaiting     NodeMaintenanceStatusType = "waiting"
+	NodeStateNew         NodeMaintenanceStatusType = "new"
+	NodeStateNewCreate   NodeMaintenanceStatusType = "new,create"
+	NodeStateNewAcquired NodeMaintenanceStatusType = "new,acquired"
+	NodeStateNewStale    NodeMaintenanceStatusType = "new,stale"
+	NodeStateNewRecreate NodeMaintenanceStatusType = "new,recreate"
+	NodeStateUpdated     NodeMaintenanceStatusType = "updated"
+	NodeStateActive      NodeMaintenanceStatusType = "active"
+	NodeStateEnded       NodeMaintenanceStatusType = "ended"
+
+	NodeStateNone NodeMaintenanceStatusType = "" // placeholder
+)
+
+type TransitionAction int
+
+const (
+	TransitionSet      TransitionAction = 0
+	TransitionInc      TransitionAction = 1
+	TransitionNoChange TransitionAction = 2
 )
 
-// Add creates a new NodeMaintenance Controller and adds it to the Manager. The Manager will set fields on the Controller
-// and Start it when the Manager is Started.
+// parse this out of NodeMaintenanceSpecAnnotation annotation
+type ReconcileNodeMaintenanceSpecInfo int32
+
+// parse this out of NodeMaintenanceStatusAnnotation annotation
+type ReconcileNodeMaintenanceStatusInfo string
+
+type ClientFactory interface {
+	createClient(config *rest.Config) (kubernetes.Interface, error)
+}
+
+// ReconcileNodeMaintenance reconciles a node object
+type ReconcileNodeMaintenance struct {
+	clientFactory ClientFactory
+	client        client.Client
+	config        *rest.Config
+
+	clientset kubernetes.Interface
+	drainer   *drain.Helper
+
+	// used only for the duration of one reconcile call. transient value.
+	// That's ok as there is one go routine that handles reconciles (here)
+	reqLogger      *log.Entry
+	node           *corev1.Node
+	specAnnoData   *ReconcileNodeMaintenanceSpecInfo
+	statusAnnoData *ReconcileNodeMaintenanceStatusInfo
```

Comment:
Data specific to an individual request is not appropriate to be stored in this structure.